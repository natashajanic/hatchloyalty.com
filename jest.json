{
  "collectCoverage": true,
  "coveragePathIgnorePatterns": ["tests"],
  "globals": {},
  "moduleFileExtensions": [
    "ts",
    "tsx",
    "js"
  ],
  "moduleNameMapper": {
    "^.+\\.(css|less|scss)$": "identity-obj-proxy",
    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/tests/assetsTransformer.ts",
    "^src/(.*)$": "<rootDir>/src/$1"
  },
  "setupFiles": [
    "raf/polyfill"
  ],
  "setupFilesAfterEnv": [
    "<rootDir>/tests/setupTests.ts"
  ],
  "snapshotSerializers": [
    "enzyme-to-json/serializer"
  ],
  "transform": {
    "\\.(ts|tsx)$": "ts-jest"
  },
  "testRegex": "/src/.*/__tests__/.*\\.(ts|tsx|js)$",
  "unmockedModulePathPatterns": [],
  "verbose": true
}
