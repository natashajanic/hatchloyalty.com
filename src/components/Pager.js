import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'gatsby'
import system from 'system-components'
import Box from '../components/Box'

const PagerWrapper = system({
  is: Box,
  display: 'flex',
  justifyContent: 'center',
  mb: 2,
  pa: 2,
})

const PagerBox = system({
  is: Box,
  alignItems: 'center',
  display: 'flex',
  height: '2em',
  justifyContent: 'center',
})

const PagerLink = system({
  is: PagerBox,
  background: 'white',
  border: 'solid 1px',
  borderColor: 'gray',
  borderRadius: '0.25em',
  color: 'green',
  minWidth: '2em',
  mx: 1,
  px: 2,
})

const PageCurrentPage = system({
  is: PagerLink,
  borderColor: 'grayDark',
  color: 'grayDark',
})

const PagerPlaceholder = system({
  is: PagerBox,
  color: 'grayDark',
  mx: 2,
})

class Pager extends React.Component {
  renderLink(pageNum, text) {
    const { pathRoot } = this.props;

    return (
      <Link
        key={`page-${pageNum}`}
        style={{ textDecoration: 'none' }}
        to={pageNum === 1 ? pathRoot : `${pathRoot}/${pageNum}`}
      >
        <PagerLink>{text}</PagerLink>
      </Link>
    )
  }

  renderPageLinks(startingPage, endingPage) {
    const { currentPage } = this.props;

    return Array(endingPage - startingPage + 1).fill(1).map((_val, index) => {
      const pageNum = (startingPage + index)
      if (pageNum === currentPage) {
        return (
          <PageCurrentPage key={`page-${pageNum}`}>
            {pageNum}
          </PageCurrentPage>
        )
      }
      return this.renderLink(pageNum, pageNum)
    });
  }

  render() {
    const { currentPage, maxPage } = this.props;

    const hasPrev = currentPage > 1
    const hasNext = currentPage < maxPage
    const startingPage = Math.max(1, currentPage - 4)
    const endingPage = Math.min(maxPage, currentPage + 4)

    return (
      <PagerWrapper>
        {hasPrev && this.renderLink(currentPage - 1, '< Prev')}
        {(startingPage > 1) && <PagerPlaceholder>...</PagerPlaceholder>}
        {this.renderPageLinks(startingPage, endingPage)}
        {(endingPage < maxPage) && <PagerPlaceholder>...</PagerPlaceholder>}
        {hasNext && this.renderLink(currentPage + 1, 'Next >')}
      </PagerWrapper>
    )
  }
}

Pager.propTypes = {
  currentPage: PropTypes.number,
  maxPage: PropTypes.number,
  pathRoot: PropTypes.string,
}

export default Pager
