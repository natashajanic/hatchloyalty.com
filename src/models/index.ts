import { FluidObject } from 'gatsby-image'

export interface IBlogPost {
  excerpt: string
  fields: {
    slug: string
  }
  frontmatter: {
    author: string
    date: string
    featuredImage?: IGraphQLImage
    tags: string[]
    title: string
  }
  html: string
}

export interface IGraphQLImage {
  childImageSharp: {
    fluid: FluidObject
  }
}

export interface IRelease {
  changes: Array<{ component: string; description: string; link?: string }>
  date: string
  name: string
}

export interface IResourcePost {
  excerpt: string
  fields: {
    slug: string
  }
  frontmatter: {
    title: string
  }
  html: string
}

export interface ITeamMember {
  headshot?: IGraphQLImage
  name: string
  position: string
}
