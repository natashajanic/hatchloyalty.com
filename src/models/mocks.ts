import { IRelease, IBlogPost, IResourcePost, ITeamMember } from '.'

export const mockBlogPost: IBlogPost = {
  excerpt: 'This is the excerpt for a blog post.',
  fields: {
    slug: '/2019-02-25-test-blog-post/',
  },
  frontmatter: {
    author: 'John Doe',
    date: 'February 25, 2019',
    featuredImage: {
      childImageSharp: {
        fluid: {
          aspectRatio: 1,
          src: 'test/imgs/test-blog-post.png',
          srcSet: '',
          sizes: '',
        },
      },
    },
    tags: ['Test', 'Press Release'],
    title: 'Test Blog Post',
  },
  html: '<div>Test blog post body content.</div>',
}

export const mockRelease: IRelease = {
  changes: [
    { component: 'Test', description: 'Test Change 1' },
    {
      component: 'Test HQ',
      description: 'Test Change 2',
      link: 'https://developer.hatchloyalty.com/test/change-2',
    },
  ],
  date: 'January 2019',
  name: 'Test Release 2019.01',
}

export const mockResourcePost: IResourcePost = {
  excerpt: 'This is the excerpt for a resource post.',
  fields: {
    slug: '/test-resource-post/',
  },
  frontmatter: {
    title: 'Test Resource Post',
  },
  html: '<div>Test resource post body content.</div>',
}

export const mockTeamMember: ITeamMember = {
  headshot: {
    childImageSharp: {
      fluid: {
        aspectRatio: 1,
        src: 'test/imgs/john-doe.png',
        srcSet: '',
        sizes: '',
      },
    },
  },
  name: 'John Doe',
  position: 'CEO',
}
