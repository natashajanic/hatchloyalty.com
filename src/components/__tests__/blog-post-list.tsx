import { shallow, ShallowWrapper } from 'enzyme'
import { Link } from 'gatsby'
import 'jest' // tslint:disable-line no-import-side-effect
import * as React from 'react'
import BlogPostPreview from 'src/components/blog-post-preview'
import Pager from 'src/components/pager'
import Text from 'src/components/text'
import { mockBlogPost } from 'src/models/mocks'
import BlogPostList, { IBlogPostListProps } from '../blog-post-list'

describe('BlogPostList', () => {
  let props: IBlogPostListProps
  let mountedComponent: ShallowWrapper<any, any, BlogPostList>

  beforeEach(() => {
    props = {
      basePath: '/tests',
      descriptionSuffix: 'tagged with "Test"',
      pageCount: 3,
      pageNum: 1,
      posts: [mockBlogPost],
      totalCount: 27,
    }
    mountedComponent = shallow(<BlogPostList {...props} />)
  })

  describe('render', () => {
    it('renders a preview for each of the provided posts', async () => {
      mountedComponent.setProps({
        posts: [
          {
            ...mockBlogPost,
            fields: { slug: '/test-post-1' },
          },
          {
            ...mockBlogPost,
            fields: { slug: '/test-post-2' },
          },
          {
            ...mockBlogPost,
            fields: { slug: '/test-post-3' },
          },
        ],
      })
      expect(mountedComponent.find(BlogPostPreview).length).toEqual(3)
      expect(
        mountedComponent
          .find(BlogPostPreview)
          .at(0)
          .prop('post').fields.slug
      ).toEqual('/test-post-1')
      expect(
        mountedComponent
          .find(BlogPostPreview)
          .at(1)
          .prop('post').fields.slug
      ).toEqual('/test-post-2')
      expect(
        mountedComponent
          .find(BlogPostPreview)
          .at(2)
          .prop('post').fields.slug
      ).toEqual('/test-post-3')
    })

    it('renders a description of the posts being shown', async () => {
      expect(mountedComponent.find(Text).length).toEqual(1)
      expect(mountedComponent.find(Text).text()).toEqual(
        'Showing 1 of 27 posts tagged with "Test"'
      )
    })

    it('renders a description with no suffix if none is specified', async () => {
      mountedComponent.setProps({ descriptionSuffix: undefined })
      expect(mountedComponent.find(Text).length).toEqual(1)
      expect(mountedComponent.find(Text).text()).toEqual(
        'Showing 1 of 27 posts '
      )
    })

    it('renders a pager if there is than 1 page worth of posts', async () => {
      expect(mountedComponent.find(Pager).length).toEqual(1)
      const pager = mountedComponent.find(Pager).at(0)
      expect(pager.prop('basePath')).toEqual(props.basePath)
      expect(pager.prop('currentPage')).toEqual(props.pageNum)
      expect(pager.prop('maxPage')).toEqual(props.pageCount)
    })

    it('omits the pager if there is only 1 page worth of posts', async () => {
      mountedComponent.setProps({ pageCount: 1 })
      expect(mountedComponent.find(Pager).length).toEqual(0)
    })

    it('matches the expected snapshot', async () => {
      expect(mountedComponent).toMatchSnapshot()
    })
  })
})
