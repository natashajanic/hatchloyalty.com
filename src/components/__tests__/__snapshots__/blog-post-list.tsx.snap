// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPostList render matches the expected snapshot 1`] = `
<Styled(Component)
  blacklist={
    Array [
      "css",
      "maxWidth",
      "m",
      "mt",
      "mr",
      "mb",
      "ml",
      "mx",
      "my",
      "p",
      "pt",
      "pr",
      "pb",
      "pl",
      "px",
      "py",
    ]
  }
  maxWidth="800px"
  mx="auto"
  px={
    Array [
      3,
      5,
    ]
  }
  py={2}
>
  <Text
    blacklist={
      Array [
        "css",
        "fontSize",
        "lineHeight",
        "m",
        "mt",
        "mr",
        "mb",
        "ml",
        "mx",
        "my",
        "p",
        "pt",
        "pr",
        "pb",
        "pl",
        "px",
        "py",
        "textStyle",
        "alignItems",
        "color",
        "bg",
        "display",
        "flex",
        "fontWeight",
        "textAlign",
      ]
    }
    fontSize={4}
    is="h2"
    lineHeight="1.5"
    m={0}
    mb={2}
    mt={4}
    textStyle="normal"
  >
    Showing 1 of 27 posts tagged with "Test"
  </Text>
  <BlogPostPreview
    key="/2019-02-25-test-blog-post/"
    post={
      Object {
        "excerpt": "This is the excerpt for a blog post.",
        "fields": Object {
          "slug": "/2019-02-25-test-blog-post/",
        },
        "frontmatter": Object {
          "author": "John Doe",
          "date": "February 25, 2019",
          "featuredImage": Object {
            "childImageSharp": Object {
              "fluid": Object {
                "aspectRatio": 1,
                "sizes": "",
                "src": "test/imgs/test-blog-post.png",
                "srcSet": "",
              },
            },
          },
          "tags": Array [
            "Test",
            "Press Release",
          ],
          "title": "Test Blog Post",
        },
        "html": "<div>Test blog post body content.</div>",
      }
    }
  />
  <Pager
    basePath="/tests"
    currentPage={1}
    maxPage={3}
  />
</Styled(Component)>
`;
